torch/lib/TensorLib/templates/Type.h

```diff
@@ -0,0 +1,63 @@
+#pragma once
+
+#include <memory>
+
+#include "TensorLib/Scalar.h"
+#include "TensorLib/ArrayRef.h"
+
+namespace tlib {
+
+class Context;
+class Storage;
+class Tensor;
+class TensorRef;
+class Generator;
+
+enum class ScalarType {
+#define DEFINE_ENUM(_1,n,_2) \
+  n,
+  TLIB_SCALAR_TYPES(DEFINE_ENUM)
+#undef DEFINE_ENUM
+  NumOptions
+};
+
+enum class Processor {
+  CPU,
+  CUDA,
+  NumOptions
+};
+
+struct CPUTag {
+  static constexpr Processor value = Processor::CPU;
+};
+struct CUDATag {
+  static constexpr Processor value = Processor::CUDA;
+};
```

Comment:
We are in the process of adding dispatch logic (eg.  `TLIB_DISPATCH_TYPE(x.type(),foo,x,prev_h);` in scalar_test.cc) that can call user-defined templates. These Tags can then be used for template specializations of those templates. This interface isn't finished yet.